lib/util/pool.ts

```diff
@@ -0,0 +1,41 @@
+import { configurationValue } from "../configuration";
+
+function concurrentDefault(): number {
+    return configurationValue<number>("pool.concurrent", 5);
+}
+
+/**
+ * Execute all provided promises with a max concurrency
+ * Results will be in the same order as the provided promises; if a promise
+ * rejects, the result will contain the error instead
+ * @param promises all promises to execute
+ * @param concurrent the max number of concurrent promise executions
+ */
+export async function executeAll<T>(promises: Array<() => Promise<T>>, concurrent: number = concurrentDefault()): Promise<T[]> {
+    let index = 0;
+    const results: any[] = [];
+    const producer = () => {
+        if (index < promises.length) {
+            const promise = promises[index]();
+            results[index] = promise;
+            index++;
+            return promise;
+        } else {
+            return null;
+        }
+    };
+
+    const PromisePool = require("es6-promise-pool");
+    const pool = new PromisePool(producer, concurrent);
+
+    pool.addEventListener("fulfilled", (event: any) => {
+        results[results.indexOf(event.data.promise)] = event.data.result;
+    });
+    pool.addEventListener("rejected", (event: any) => {
+        results[results.indexOf(event.data.promise)] = event.data.error;
```

Comment:
`Promise.all` rejects immediately on any rejected promise. Are you intentionally choosing different behavior?

Reply:
I think I’m our usecaes you want to actually exit early. Keep going and collect. Imho that makes more sense. 

Maybe I should add the `Promise<Array<T | Error>>` as return value. 

What you recommend?

Reply:
Turns out the `es6-promise-pool` rejects immediately on any rejected promise. So we are not changing the semantics here.

Actually I don't like it that much but that is for another day.